File: packages/python/sheet_compressor/encodings/chart_descriptors.py
"""Chart-descriptor rendering (SPEC §6)."""

from typing import Any, List, Mapping, Optional


def _escape_quoted(s: str) -> str:
    """SPEC §6.1: double-quoted token field (title, xAxis, yAxis)."""
    return (
        s.replace("\\", "\\\\")
        .replace('"', '\\"')
        .replace("\n", "\\n")
        .replace("\r", "\\r")
        .replace("\t", "\\t")
    )


def _escape_series_name(s: str) -> str:
    """SPEC §6.1: a single series name inside `series=[…]`."""
    return (
        s.replace("\\", "\\\\")
        .replace(",", "\\,")
        .replace("]", "\\]")
        .replace("\n", "\\n")
        .replace("\r", "\\r")
        .replace("\t", "\\t")
    )


def render_chart_token(chart: Mapping[str, Any]) -> str:
    """SPEC §6.1: render one descriptor. `name` is NOT rendered."""
    parts: List[str] = [f"CHART({chart['type']})@{chart['anchorRange']}"]
    title = chart.get("title")
    if title is not None:
        parts.append(f'title="{_escape_quoted(title)}"')
    data_ranges = chart.get("dataRanges")
    if data_ranges:
        parts.append(f"data={','.join(data_ranges)}")
    series = chart.get("series")
    if series:
        parts.append("series=[" + ",".join(_escape_series_name(s) for s in series) + "]")
    axes = chart.get("axes") or {}
    x_axis = axes.get("x")
    if x_axis is not None:
        parts.append(f'xAxis="{_escape_quoted(x_axis)}"')
    y_axis = axes.get("y")
    if y_axis is not None:
        parts.append(f'yAxis="{_escape_quoted(y_axis)}"')
    return " ".join(parts)


def render_chart_block(charts: Optional[List[Mapping[str, Any]]]) -> str:
    """SPEC §6.2: tokens joined by `\\n` in input order. `""` when missing/empty."""
    if not charts:
        return ""
    return "\n".join(render_chart_token(c) for c in charts)


def append_chart_block(cell_string: str, chart_block: str) -> str:
    """SPEC §6.2: combine an encoding's cell-string with the chart block."""
    if chart_block == "":
        return cell_string
    if cell_string == "":
        return chart_block
    return f"{cell_string}\n{chart_block}"

File: packages/typescript/src/encodings/chartDescriptors.ts
import type { ChartDescriptor } from "../types.ts";

/**
 * Per SPEC §6.1: the contents of a double-quoted token field (title, xAxis,
 * yAxis). Backslash first so later rules' backslashes aren't double-escaped,
 * then the quote, then the whitespace controls.
 */
function escapeQuoted(s: string): string {
  return s
    .replace(/\\/g, "\\\\")
    .replace(/"/g, '\\"')
    .replace(/\n/g, "\\n")
    .replace(/\r/g, "\\r")
    .replace(/\t/g, "\\t");
}

/**
 * Per SPEC §6.1: a single series name inside `series=[…]`. Backslash first,
 * then the bracket-list delimiters (`,` and `]`), then the whitespace controls.
 */
function escapeSeriesName(s: string): string {
  return s
    .replace(/\\/g, "\\\\")
    .replace(/,/g, "\\,")
    .replace(/\]/g, "\\]")
    .replace(/\n/g, "\\n")
    .replace(/\r/g, "\\r")
    .replace(/\t/g, "\\t");
}

/**
 * Render a single chart descriptor to the SPEC §6.1 token form. Optional
 * fields are omitted entirely when undefined or (for `dataRanges`/`series`)
 * when the source array is empty. The `name` field is intentionally not
 * rendered — it is a developer-facing identifier, not LLM context.
 */
export function renderChartToken(chart: ChartDescriptor): string {
  const parts: string[] = [`CHART(${chart.type})@${chart.anchorRange}`];
  if (chart.title !== undefined) {
    parts.push(`title="${escapeQuoted(chart.title)}"`);
  }
  if (chart.dataRanges && chart.dataRanges.length > 0) {
    parts.push(`data=${chart.dataRanges.join(",")}`);
  }
  if (chart.series && chart.series.length > 0) {
    parts.push(`series=[${chart.series.map(escapeSeriesName).join(",")}]`);
  }
  if (chart.axes?.x !== undefined) {
    parts.push(`xAxis="${escapeQuoted(chart.axes.x)}"`);
  }
  if (chart.axes?.y !== undefined) {
    parts.push(`yAxis="${escapeQuoted(chart.axes.y)}"`);
  }
  return parts.join(" ");
}

/**
 * Per SPEC §6.2: tokens joined by `\n` in input order, no trailing newline.
 * Returns `""` when `charts` is missing or empty.
 */
export function renderChartBlock(
  charts: ChartDescriptor[] | undefined,
): string {
  if (!charts || charts.length === 0) return "";
  return charts.map(renderChartToken).join("\n");
}

/**
 * Per SPEC §6.2: append the chart block to a cell-string with the documented
 * separator rule. Empty inputs collapse cleanly: empty cells + charts → charts
 * only; cells + no charts → cells only; both empty → empty string.
 */
export function appendChartBlock(cellString: string, chartBlock: string): string {
  if (chartBlock === "") return cellString;
  if (cellString === "") return chartBlock;
  return `${cellString}\n${chartBlock}`;
}
